### ISSUE TITLE
[Feature Request] MadTracker 2 and Extreme Tracker module formats.

### ISSUE THREAD

**Issue by AliceLR** [libxmp/libxmp#379]
The MadTracker 2 (MT2) and Extreme Tracker (AMS) formats are supported by libmodplug but not by libxmp, as far as I can tell. Aside from the DSIK internal format (#120), these are the only sane libmodplug formats that libxmp does not support at all. (libmodplug also supports X-Tracker, which has an existing libxmp loader that needs to be fixed.)

**Comment by sagamusix** [comment]
One of the "core" features of the MT2 format is stereo sample support and libxmp still doesn't support that. I'd urge not to add any MT2 support until libxmp finally has stereo sample support because otherwise this means there will be yet another feature-incomplete implementation of this format. (I'm particularly grumpy about this one because I make a lot of use of stereo samples in my tracks and several widespread module decoders don't support that, so a considerable amount of peopel will never hear them "as intended". Trackers have supported stereo samples for more than 20 years, it's hardly a modern and rare feature at this point.)

MT2 also supports several DSPs which are currently not supported by any implementation outside of MT2 itself, I got partial source code (x86 assembly) of those but it's difficult to get them to work without all the supporting source around it.

**Comment by AliceLR** [comment]
I agree that stereo support would generally be a very good thing to have, and that's generally something I'd be interested in looking into fixing. (I've used them from time to time as well, though I don't have stats; many of my large modules only distributed as OGGs were lost in a drive crash.)

I don't know how libopenmpt's support for these formats is but the bar I was setting is "better support than libmodplug"... in the case of MT2, I think that means not having trash and/or nonexistent bounds checking and supporting any of the effects at all, though. 🙄

**Comment by sagamusix** [comment]
That bar isn't too high then... My point is, I think having even more incomplete implementations of formats isn't going to help anyone, because it just inevitably means that even more people will be exposed to those incomplete implementations, and they will have to compare even more libraries and input plugins against each other to figure out which one does a slightly better job than the other. It's one of the reasons why libxmp removed support for Adlib formats after all.

OpenMPT's MT2 support is certainly much better than libmodplug's, but some advanced features (like the built-in DSPs) are sadly missing and not very practical to implement. AMS support is better than libmodplug's as well, and that should be fairly straightforward to add to libxmp (except maybe for the higher-resolution envelopes, in particular the pitch envelope).

**Comment by AliceLR** [comment]
> My point is, I think having even more incomplete implementations of formats isn't going to help anyone, because it just inevitably means that even more people will be exposed to those incomplete implementations, and they will have to compare even more libraries and input plugins against each other to figure out which one does a slightly better job than the other. It's one of the reasons why libxmp removed support for Adlib formats after all.

I agree, and don't get me wrong. I'm not advocating for adding broken/partial format support. I also have no intention of implementing these right now, as fixing existing formats is a much better use of time than addressing this request.

I think any feature libmodplug can claim to have that other libraries don't can potentially an argument against switching away from libmodplug, which still has no shortage of potential CVEs (https://github.com/Konstanty/libmodplug/pull/58), so I think *correct* support these formats (and DMF, and DSIK) should be implemented at some point.

**Comment by sagamusix** [comment]
I think the main problem is that the (few?) projects that are stuck with libmodplug just have no motivation or resources to switch *despite* the security issues being well-known. The lack of an alternative doesn't seem to be the main motivation for most of them. I know VLC is still using it because for them libopenmpt was "too big" when we last talked (which is mostly due to heavier use of the C++ standard library). Luckily on Debian and derivates our libopenmpt-modplug emulation layer is used to replace libmodplug, which at least avoids that attack surface on some well-known media applications on many Linux distributions.
